## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

/var/folders/g6/0bc2_tl1389c2rmm69jgtcr00000gn/T/ipykernel_86266/2050583822.py:14: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)
/var/folders/g6/0bc2_tl1389c2rmm69jgtcr00000gn/T/ipykernel_86266/2050583822.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
df = contrib
df.shape

(1125659, 18)

- **1a answer here:** Yes, seems to be a reasonable amount of rows and the correct amount of columns 

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
df.columns

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')

- **1b answer here:** Yes

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
df.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:** cand_id seems to be in the form of alphabet+8*number; cand_nm seems to be standard across rows as well, corresponding to the id; contbr_st seems to be the state's abbreviated form and is correct for the top 5 rows as all the locations exist in the state

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
df["election_tp"]

0          P2016
1          P2016
2          P2016
3          P2016
4          P2016
           ...  
1125654    P2016
1125655    P2016
1125656    P2016
1125657    P2016
1125658    P2016
Name: election_tp, Length: 1125659, dtype: object

- **1d answer here:** election_tp is the election type, so whether it's primary, general, convention, runoff, special, recount, or other election, also seems to match the PYYYY format in the docs

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
df.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** contbr_zip: object; contb_receipt_amt: float64; contb_receipt_dt: datetime64[ns]

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
df.count().sort_values(ascending=False)

cmte_id              1125659
cand_id              1125659
tran_id              1125659
file_num             1125659
form_tp              1125659
contb_receipt_dt     1125659
contb_receipt_amt    1125659
contbr_st            1125659
contbr_nm            1125659
cand_nm              1125659
contbr_city          1125633
contbr_zip           1125564
election_tp          1124234
contbr_occupation    1115260
contbr_employer       967757
memo_text             501148
memo_cd               144268
receipt_desc           15045
dtype: int64

- **1f answer here:** most non-nulls: cmte_id, cand_id, tran_id, file_num, form_tp, contb_receipt_dt, contb_receipt_amt, contbr_st, contbr_nm; I would not recommend dropping a column just because of nulls, instead the null value should be accounted for in the data analysis. If you had to drop columns, memo_cd and recept_desc are both very small percentages of the total dataset, so I would drop those.

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE
cand_id_groups = df.groupby("cand_id")

def check_all_same(group, column):
    return (group[column].nunique() == 1)

result = cand_id_groups.apply(lambda x: check_all_same(x, "cand_nm"))
print(result)

bernie_group = cand_id_groups.get_group("P60007168")
(bernie_group["cand_nm"] == "Sanders, Bernard").all()

cand_id
P00003392    True
P20002671    True
P20002721    True
P20003281    True
P20003984    True
P40003576    True
P60003670    True
P60005915    True
P60006046    True
P60006111    True
P60006723    True
P60007168    True
P60007242    True
P60007572    True
P60007671    True
P60007697    True
P60008059    True
P60008398    True
P60008521    True
P60008885    True
P60009685    True
P60022654    True
P80001571    True
P80003379    True
P80003478    True
dtype: bool


/var/folders/g6/0bc2_tl1389c2rmm69jgtcr00000gn/T/ipykernel_86266/196516410.py:7: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result = cand_id_groups.apply(lambda x: check_all_same(x, "cand_nm"))


np.True_

- **1g answer here:** No, there do not seem to be any issues with matching cand_nm to cand_id

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
(df["contbr_st"] == "CA").all()

np.True_

- **1h answer here:** No

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
duplicates = df["tran_id"].duplicated()
print(duplicates.sum())
print(df[duplicates])
# print(df[duplicates].sort_values(by="contb_receipt_dt"))

3454
           cmte_id    cand_id                    cand_nm          contbr_nm  \
15650    C00574624  P60006111  Cruz, Rafael Edward 'Ted'    HAWKINS, THOMAS   
16641    C00574624  P60006111  Cruz, Rafael Edward 'Ted'      KREMER, DEBRA   
16963    C00574624  P60006111  Cruz, Rafael Edward 'Ted'    GOODWIN, HOWARD   
17290    C00574624  P60006111  Cruz, Rafael Edward 'Ted'   VALKO, BONNIE L.   
17960    C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PICKERING, VICTOR   
...            ...        ...                        ...                ...   
1125589  C00573519  P60005915        Carson, Benjamin S.      LONGWELL, SAM   
1125596  C00573519  P60005915        Carson, Benjamin S.      MANN, VIVIANE   
1125597  C00573519  P60005915        Carson, Benjamin S.      MANN, VIVIANE   
1125598  C00573519  P60005915        Carson, Benjamin S.      MANN, VIVIANE   
1125614  C00573519  P60005915        Carson, Benjamin S.    HANSEN, WILLIAM   

           contbr_city contbr_st     contbr_zi

- **1i answer here:** 3454 duplicate entries. I can't see any pattern here, I thought maybe all the contb_receipt_amt values were divisible by 25, but that is not the case. I also checked whether there were any patterns with the dates, but they seem to be reasonable as well.

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE
print((df["contb_receipt_amt"] < 0).sum())
df[(df["contb_receipt_amt"] < 0)]

11896


cmte_id    cand_id                    cand_nm  \
19       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
23       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
81       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
190      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
213      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125008  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125317  C00580399  P60008521   Christie, Christopher J.   
1125427  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125446  C00573519  P60005915        Carson, Benjamin S.   
1125582  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   

                    contbr_nm             contbr_city contbr_st  \
19           JOLLIFF, RICHARD                   CHICO        CA   
23           JOLLIFF, RICHARD                   CHICO        CA   
81           JOLLIFF, RICHARD                   CHICO        CA   
190          NOWELL, DIANA L.  RANCHO SANTA MARGARITA        CA   
213        LICHTY, ANDREW MR.               SAN DIEGO        CA   
...                       ...                     ...       ...   
1125008  NELSON, PETER C. MR.         SAN LUIS OBISPO        CA   
1125317       ASCHER, STEPHEN                PASADENA        CA   
1125427       HANSEN, WILLIAM                GLENDALE        CA   
1125446        PECK, JOHN JR.         RANCHO SANTA FE        CA   
1125582        FARRELL, PETER                LA JOLLA        CA   

            contbr_zip                     contbr_employer  \
19      959,289,507.00                       SELF EMPLOYED   
23      959,289,507.00                       SELF EMPLOYED   
81      959,289,507.00                       SELF EMPLOYED   
190     926,884,928.00  CAPISTRAND UNIFIED SCHOOL DISTRICT   
213     921,096,720.00                       SELF EMPLOYED   
...                ...                                 ...   
1125008      934018000                       SELF EMPLOYED   
1125317      911013113                             MIRAMAR   
1125427      912081507                             RETIRED   
1125446      920670829                    PECK ENTERPRISES   
1125582      920375633                              RESMED   

            contbr_occupation  contb_receipt_amt contb_receipt_dt  \
19                    RANCHER             -25.00       2016-04-29   
23                    RANCHER            -150.00       2016-04-29   
81                    RANCHER             -60.00       2016-04-14   
190      LIBRARIAN TECHNICIAN            -100.00       2016-04-11   
213               REAL ESTATE             -25.00       2016-04-30   
...                       ...                ...              ...   
1125008               DENTIST          -2,700.00       2015-07-31   
1125317             EXECUTIVE          -2,700.00       2016-01-30   
1125427               RETIRED          -5,400.00       2015-07-01   
1125446             PRESIDENT          -2,700.00       2016-01-01   
1125582              CHAIRMAN          -2,700.00       2015-08-28   

                     receipt_desc memo_cd                 memo_text form_tp  \
19       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
23       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
81       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
190      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
213      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
...                           ...     ...                       ...     ...   
1125008  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125317   REATTRIBUTION TO SPOUSE       X   REATTRIBUTION TO SPOUSE   SA17A   
1125427                       NaN     NaN              CHARGED BACK   SA17A   
1125446  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125582  REDESIGNATION TO GENERAL       X 

- **1j answer here:** 11896 negative donations. Yeah a lot of these are "redesignated to general" or "charged back" in the memo_txt field, implying that the donation was either refunded or redirected, so money was taken from the account

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE
start_date = pd.to_datetime('2014-01-01')
end_date = pd.to_datetime('2023-06-07')

df[(df['contb_receipt_dt'] < start_date) | (df['contb_receipt_dt'] > end_date)]

cmte_id    cand_id       cand_nm          contbr_nm contbr_city  \
9932  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   
9994  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   

     contbr_st     contbr_zip contbr_employer  contbr_occupation  \
9932        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   
9994        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   

      contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
9932             -20.00       2013-11-05          NaN       X   
9994              20.00       2013-11-05          NaN       X   

                        memo_text form_tp  file_num             tran_id  \
9932  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.2.0615   
9994  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.3.0615   

     election_tp  
9932       P2016  
9994       G2016

- **1k answer here:** Yes, two seem to be on November 5th, 2013, and they are well formatted for the analysis.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:** Yes

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:** Yes

**1l.3** What columns are important?

- **1l.3 answer here:** Depends on the questions you are trying to answer. So far, cand_id and cand_nm are the most significant.

**1l.4** What columns can be dropped?

- **1l.4 answer here:** Ideally none, but memo_cd and receipt_desc are both columns with very high percentages of nulls, so they can be dropped.

**1l.5** What are the data problems?

- **1l.5 answer here:** High percentage of nulls in memo_cd and receipt_desc. Around 3.5k duplicate rows. 

**1l.6** List any assumptions so far:

- **1l.6 answer here:** Data is valid and legitimately pulled from election data. 

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [13]:
# 2a YOUR CODE HERE
df = df[df['election_tp'].str[0] == 'P']
df.shape

(810488, 18)

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [14]:
# 2b YOUR CODE HERE
df = df[df["cand_nm"] == "Sanders, Bernard"]
df.shape

(407171, 18)

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [15]:
# 2c YOUR CODE HERE
five_digit_zips = df[df['contbr_zip'].astype(str).str.match(r'^\d{5}$')]
five_digit_zips["zipcode"] = five_digit_zips["contbr_zip"]
df["zipcode"] = None
df.update(five_digit_zips)
df = df[df["zipcode"].notnull()]
df.shape

/var/folders/g6/0bc2_tl1389c2rmm69jgtcr00000gn/T/ipykernel_86266/3839195870.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  five_digit_zips["zipcode"] = five_digit_zips["contbr_zip"]


(2197, 19)

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [16]:
# 2d YOUR CODE HERE
df = df[df["contb_receipt_amt"] >= 0]
df.shape

(2193, 19)

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [17]:
# 2e YOUR CODE HERE
df = df[["contb_receipt_amt", "zipcode", "contb_receipt_dt"]]
df.shape

(2193, 3)

- **2e answer here:** I do not need anything other columns for the following two questions.

**2f.** List any assumptions that you made up to this point:

- **2f answer here:** Data is valid and legitimately pulled from election data. 

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [18]:
# 3a YOUR CODE HERE
zipcode_groups = df.groupby("zipcode")
greatest_contb_amt = zipcode_groups["contb_receipt_amt"].sum().sort_values(ascending=False)
print(greatest_contb_amt[0:1])
highest_contb_count = zipcode_groups.size().sort_values(ascending=False)
highest_contb_count[0:1]

zipcode
94301   3,200.00
Name: contb_receipt_amt, dtype: float64


zipcode
94602    75
dtype: int64

- **3a answer here:** Highest Count of Contributions: 94602, Most Dollar Amount: 94301

**3b.** What day(s) of the month do most people donate? (10 points)

In [19]:
# 3b YOUR CODE HERE
day_group = df.groupby(df['contb_receipt_dt'].dt.day)
day_group.size().sort_values(ascending=False)[0:1]

contb_receipt_dt
29    209
dtype: int64

- **3b answer here:** Seems to be the 29th of the month. LINK TO REPO: https://github.com/ajha7/DATASCI200-Assignments

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k